### PR TITLE
Duplication of companion methods after using FromTasty fixed

### PR DESCRIPTION
**Pull request by VladimirNik**
Fix for: https://github.com/lampepfl/dotty/issues/977

The problem was due to `scalacLinkedClass` method invocation during unpickling. Companion methods should be generated only if symbol returned by `scalacLinkedClass` is among unpickled symbols.

Review by @odersky, @DarkDimius

### CODE REVIEW

File: src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 1]
```diff
@@ -590,7 +590,7 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table) {
         case TYPEDEF | TYPEPARAM =>
           if (sym.isClass) {
             val companion = sym.scalacLinkedClass
-            if (companion != NoSymbol) {
+            if (companion != NoSymbol && symAtAddr.valuesIterator.contains(companion)) {
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: why not `symAtAddr.contains(companion)`?

Comment (VladimirNik) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: @DarkDimius because symAtAddr is a map, key is an address.

Comment (DarkDimius) [inline src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: `valuesIterator` takes linear time to check if element is contained.
It means that this change makes reading tasty quadratic in number of `TypeDefs`. 

if `companion != NoSymbol`, where does `companion` originate from if not from Tasty?

**Comment by DarkDimius** [comment]
LGTM

**Pull request merged**